### ISSUE TITLE
eval on fintuning model

### ISSUE THREAD

**Issue by liuao743** [magic-research/PLLaVA#50]
I do full fintuning on all layers of the network, including the vision encoder, projection layer, and language model. I don't use deepspeed. In the checkpoint_dir, I get .pt files as follows:
![1716519539416](https://github.com/magic-research/PLLaVA/assets/62134958/75f070ed-455e-4821-8a23-d5c76d5c96e2)
 How can I use the my after-training weight ?
@ermu2001

**Comment by liuao743** [comment]
> Hi,
> 
> ckpt_epoch** was to be loaded by accelerate for resuming. The model saved that directly fits into evaluation would be in the folders starting with "pretrained_*xx" (xx are digits and * is step/epoch). Do you see that in the result directory?

Of course, before I use deepspeed, the directory "pretrained_xx" are as follows:
![5cb7cd2b315004c840580a604d61714](https://github.com/magic-research/PLLaVA/assets/62134958/f2ce5aac-af88-4a6d-bae9-db4c099986f3)
 
when I fintuning it useing deepspeed, they are as follows:
![b18012cb0f53455a30deeda60a40058](https://github.com/magic-research/PLLaVA/assets/62134958/54ed0472-3d1e-4b0a-877c-07b3219ecfe6)

Now I can use it inference. But when I merge them into one directory as follows:
![62eab11584412f7ef407c4c7cc7a241](https://github.com/magic-research/PLLaVA/assets/62134958/814f3f61-f7fb-4ad5-9e0d-d06332cc573e)
and set it to weight_dir, model_dir, the inference result always repeat one sentence for all input.
When I set it to weight_dir, but set the origin "llava-v1.6-vicuna-7b-hf" to model_dir, It can output correctly.
But I do full fintuning on all layers of the network, including the vision encoder, projection layer, and language model, It seems that the full finetuning of the language model is so bad, have you ever encountered this situation before ?

**Comment by ermu2001** [comment]
Did you tried to load weights from pllava (MODELS/pllava-7b) and do continual training with deepspeed zero3?

During investigation into this. I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. Are there any suspicious log output indicating faults in the training process. This is should be fixed at [dev](https://github.com/magic-research/PLLaVA/tree/dev) branch now.

We did not unfrozen all three parts of the model. Unfreezing setting is yet to be discovered. However, I don't thinking this video training would result in total nonsense response. I've unfreeze the language model and it seems that it is inferior to training with lora. I'll probably run a experiments unfreezing all counterparts tomorrow. I'll I provide an update on the progress at that time.

You can probably checkout dev branch and see whether it remains the same.

**Comment by liuao743** [comment]
> Did you tried to load weights from pllava (MODELS/pllava-7b) and do continual training with deepspeed zero3?
> 
> During investigation into this. I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. Are there any suspicious log output indicating faults in the training process. This is should be fixed at [dev](https://github.com/magic-research/PLLaVA/tree/dev) branch now.
> 
> We did not unfrozen all three parts of the model. Unfreezing setting is yet to be discovered. However, I don't thinking this video training would result in total nonsense response. I've unfreeze the language model and it seems that it is inferior to training with lora. I'll probably run a experiments unfreezing all counterparts tomorrow. I'll I provide an update on the progress at that time.
> 
> You can probably checkout dev branch and see whether it remains the same.

Firstly, I used accelerate, but did not use deepspeed. In this case, the loss has converged, but the saved weights are not available (as I mentioned earlier). So I continued to use deepspeed training on the basis of the new checkpoint, and encountered the issues mentioned above. It may be due to “I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. ”
In my training process, I didn't get ang error.
Also, I get another problem, I have saved the learning rate and loss during the training phase and created a curve. I found that the loss is normal, but the learning rate is very strange. I conducted a thorough investigation of the source code and printed the scheduler, but did not find any problems. If possible, I hope you will also pay attention to this issue when trying. P. S. My training phase configuration is: two machines, each with 8 80G A100 sheets.
![image](https://github.com/magic-research/PLLaVA/assets/62134958/828ad7ca-f08a-44b3-9240-71e8538c3dd6)

**Comment by liuao743** [comment]
> I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method

And could you please explain this sentence in more detail：“ I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method”. I didn't get any error when I do resuming training.

**Comment by liuao743** [comment]
> Did you tried to load weights from pllava (MODELS/pllava-7b) and do continual training with deepspeed zero3?
> 
> During investigation into this. I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. Are there any suspicious log output indicating faults in the training process. This is should be fixed at [dev](https://github.com/magic-research/PLLaVA/tree/dev) branch now.
> 
> We did not unfrozen all three parts of the model. Unfreezing setting is yet to be discovered. However, I don't thinking this video training would result in total nonsense response. I've unfreeze the language model and it seems that it is inferior to training with lora. I'll probably run a experiments unfreezing all counterparts tomorrow. I'll I provide an update on the progress at that time.
> 
> You can probably checkout dev branch and see whether it remains the same.

Today I trained the model from the origin model_dir, using deepspeed2. "result in total nonsense response" still appearing.

**Comment by ermu2001** [comment]
> > Did you tried to load weights from pllava (MODELS/pllava-7b) and do continual training with deepspeed zero3?
> > During investigation into this. I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. Are there any suspicious log output indicating faults in the training process. This is should be fixed at [dev](https://github.com/magic-research/PLLaVA/tree/dev) branch now.
> > We did not unfrozen all three parts of the model. Unfreezing setting is yet to be discovered. However, I don't thinking this video training would result in total nonsense response. I've unfreeze the language model and it seems that it is inferior to training with lora. I'll probably run a experiments unfreezing all counterparts tomorrow. I'll I provide an update on the progress at that time.
> > You can probably checkout dev branch and see whether it remains the same.
> 
> Firstly, I used accelerate, but did not use deepspeed. In this case, the loss has converged, but the saved weights are not available (as I mentioned earlier). So I continued to use deepspeed training on the basis of the new checkpoint, and encountered the issues mentioned above. It may be due to “I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. ” In my training process, I didn't get ang error. Also, I get another problem, I have saved the learning rate and loss during the training phase and created a curve. I found that the loss is normal, but the learning rate is very strange. I conducted a thorough investigation of the source code and printed the scheduler, but did not find any problems. If possible, I hope you will also pay attention to this issue when trying. P. S. My training phase configuration is: two machines, each with 8 80G A100 sheets. ![image](https://private-user-images.githubusercontent.com/62134958/334347308-828ad7ca-f08a-44b3-9240-71e8538c3dd6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5NTU5NTUsIm5iZiI6MTcxNjk1NTY1NSwicGF0aCI6Ii82MjEzNDk1OC8zMzQzNDczMDgtODI4YWQ3Y2EtZjA4YS00NGIzLTkyNDAtNzFlODUzOGMzZGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDA0MDczNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiYjc1NTQyNTIzMjgwZWUwZTUxY2QzYzliZTEzNTdhNTcxYWUwYmIwYjM5MDIxODY2ZGQ4NDU2OGExZmQyODkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iL15_r_Q0qbF8cfbOSpq8ZeqkRPT2KwCoo8kkikbuDI)

The loss is too high, this pretty much indicates the pretrained models' weighs are not loaded. As an example, several training curve looks like this, where it should be around 1-2 as the model is fully trained before starting training with videos:
![image](https://github.com/magic-research/PLLaVA/assets/55656210/8d3c04a0-a99b-4d27-90ec-aeda53266115)

# how to load weights using deepspeed
Deepspeed cut up the model on different processes, such that requires code to gather the weights then load then dispatch again as in:
https://github.com/magic-research/PLLaVA/blob/9a03284c7f31ea5787acf2dbf868e32a0692f7fe/tasks/model_utils.py#L28-L32
This is now supported in the dev branch.

**Comment by liuao743** [comment]
> > > Did you tried to load weights from pllava (MODELS/pllava-7b) and do continual training with deepspeed zero3?
> > > During investigation into this. I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. Are there any suspicious log output indicating faults in the training process. This is should be fixed at [dev](https://github.com/magic-research/PLLaVA/tree/dev) branch now.
> > > We did not unfrozen all three parts of the model. Unfreezing setting is yet to be discovered. However, I don't thinking this video training would result in total nonsense response. I've unfreeze the language model and it seems that it is inferior to training with lora. I'll probably run a experiments unfreezing all counterparts tomorrow. I'll I provide an update on the progress at that time.
> > > You can probably checkout dev branch and see whether it remains the same.
> > 
> > 
> > Firstly, I used accelerate, but did not use deepspeed. In this case, the loss has converged, but the saved weights are not available (as I mentioned earlier). So I continued to use deepspeed training on the basis of the new checkpoint, and encountered the issues mentioned above. It may be due to “I discovered that the main code now doesn't support directly loading weights from existing pretrained model except from the from_pretrained method. ” In my training process, I didn't get ang error. Also, I get another problem, I have saved the learning rate and loss during the training phase and created a curve. I found that the loss is normal, but the learning rate is very strange. I conducted a thorough investigation of the source code and printed the scheduler, but did not find any problems. If possible, I hope you will also pay attention to this issue when trying. P. S. My training phase configuration is: two machines, each with 8 80G A100 sheets. ![image](https://private-user-images.githubusercontent.com/62134958/334347308-828ad7ca-f08a-44b3-9240-71e8538c3dd6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5NTU5NTUsIm5iZiI6MTcxNjk1NTY1NSwicGF0aCI6Ii82MjEzNDk1OC8zMzQzNDczMDgtODI4YWQ3Y2EtZjA4YS00NGIzLTkyNDAtNzFlODUzOGMzZGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDA0MDczNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiYjc1NTQyNTIzMjgwZWUwZTUxY2QzYzliZTEzNTdhNTcxYWUwYmIwYjM5MDIxODY2ZGQ4NDU2OGExZmQyODkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iL15_r_Q0qbF8cfbOSpq8ZeqkRPT2KwCoo8kkikbuDI)
> 
> The loss is too high, this pretty much indicates the pretrained models' weighs are not loaded. As an example, several training curve looks like this, where it should be around 1-2 as the model is fully trained before starting training with videos: ![image](https://private-user-images.githubusercontent.com/55656210/334646698-8d3c04a0-a99b-4d27-90ec-aeda53266115.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5NjExMzgsIm5iZiI6MTcxNjk2MDgzOCwicGF0aCI6Ii81NTY1NjIxMC8zMzQ2NDY2OTgtOGQzYzA0YTAtYTk5Yi00ZDI3LTkwZWMtYWVkYTUzMjY2MTE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDA1MzM1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhODc1NDk4OGQ4ZTUyNmMxM2E1ZTg2YTU2YjFiZTVlY2RhMjYwMDA3NmQ3OTBlNWQ1YzRhZDUzNWE0MDI1ODkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.g6RjgowEios-7wE7n1JgSibVfE8vU0HwvLTWH3OChdM)
> 
> # how to load weights using deepspeed
> Deepspeed cut up the model on different processes, such that requires code to gather the weights then load then dispatch again as in:
> 
> https://github.com/magic-research/PLLaVA/blob/9a03284c7f31ea5787acf2dbf868e32a0692f7fe/tasks/model_utils.py#L28-L32
> 
> 
> This is now supported in the dev branch. So if you want to do continual training on pllava, I suggest you switch to dev or wait for a few days for us to finalize the changes. Otherwise you can try start training with llava-hf/llava-v1.6-vicuna-7b-hf

Yeah, I think this must be my mistake, I will make further modifications and try new experiments. Thank you very much!
